### PR TITLE
Make url hash a base32 encode

### LINKED ISSUE #1902: [BUG] Short url can end in a - (dash). Not all url parses can handle this, resulting in broken links.
**Describe the bug**
I guess see title

**To Reproduce**
Click share, short url, hope for a dash at the end

**Expected behavior**
A common parseable url without a dash at the end

**Link**
https://godbolt.org/z/PiycZ-

**Screenshots**
- N/A

**Desktop (please complete the following information):**
- N/A

**Additional context**
- N/A

### PR DESCRIPTION
**Pull request by RabsRincon**
Avoids most of the problems present with the current base64 implementation
Closes #1902

### CODE REVIEW

File: lib/utils.js [hunk 1]
```diff
@@ -221,3 +221,46 @@ exports.replaceAll = function replaceAll(line, oldValue, newValue) {
...
+        const character = BASE32_ALPHABET[word];
+        output += character;
+        bits -= 5;
+        // Shift out the newly processed word
+        digest = (digest >>> 5);
+    }
+
+    for (const byte of buffer) {
+        // Append current byte to digest
+        digest = ((byte) * Math.pow(2, bits)) | digest;
```
Comment (RabsRincon) [inline lib/utils.js:252]: For my future self: `digest = (byte << bits) | digest`

File: lib/storage/storage.js [hunk 2]
```diff
@@ -24,15 +24,17 @@
...
     /***
      * @type {string[]}
      */
     profanities = require('profanities');
 
-const FILE_HASH_VERSION = 'Compiler Explorer Config Hasher';
-const USABLE_NAME_MAX_LENGTH = 9; // Quite generous
-const ILLEGIBLE_LETERS = ['l', 'I', 'O', '0', '1']; // L i and o are allowed. Readable if the others are not present
+const FILE_HASH_VERSION = 'Compiler Explorer Config Hasher 2';
+/* How long a string to check for possible unusbale hashes (Profanities or confusing text)
```
Comment (mattgodbolt) [inline lib/storage/storage.js:34]: Typo! "unusable" :)

File: lib/utils.js (hunk 1 above)
Comment (mattgodbolt) [reply lib/utils.js:252]: Why can't we use the `(byte << bits)`? What am I forgetting about javascript here haha

File: lib/utils.js [hunk 3]
```diff
@@ -221,3 +221,46 @@ exports.replaceAll = function replaceAll(line, oldValue, newValue) {
     }
     return line;
 };
+
+// Initally based on http://philzimmermann.com/docs/human-oriented-base-32-encoding.txt
```
Comment (mattgodbolt) [inline lib/utils.js:225]: :heart:

File: test/utils-tests.js [hunk 4]
```diff
@@ -349,4 +349,21 @@ describe('replaces all substrings', () => {
...
-});
\ No newline at end of file
+});
+
+describe('encodes in our version of base32', () => {
+    function doTest(original, expected) {
+        utils.base32Encode(Buffer.from(original)).should.equal(expected);
+    }
+
+   it('returns what we expect', () => {
```
Comment (mattgodbolt) [inline test/utils-tests.js:359]: Do these test the case where we need the "last pass" ?

File: lib/utils.js (hunk 1 above)
Comment (RabsRincon) [reply lib/utils.js:252]: No reason not to! I even left a note to myself an hour ago to remind me tomorrow to switch it!

File: test/utils-tests.js (hunk 4 above)
Comment (RabsRincon) [reply test/utils-tests.js:359]: Yes! But I just made it more explicit on the tests now

**Pull request merged**